RecoVertex/VertexPrimitives/src/BasicSingleVertexState.cc

```diff
@@ -1,90 +1,244 @@
+                                               const double & weightInMix)
+  : thePos(pos), 
+    theTime(time), 
+    theErr(posTimeErr), 
+    theWeight(AlgebraicSymMatrix44()), 
+    theWeightTimesPos(AlgebraicVector4()), 
+    theWeightInMix(weightInMix),
+    thePosAvailable(true), theTimeAvailable(true), theErrAvailable(true), theWeightAvailable(false), theWeightTimesPosAvailable(false),
+    valid(true), vertexIs4D(true)
+{}
+
+
+BasicSingleVertexState::BasicSingleVertexState(const GlobalPoint & pos,
+                                               const double time,
+                                               const GlobalWeight & posTimeWeight,
+                                               const double & weightInMix)
+  : thePos(pos), 
+    theTime(time), 
+    theErr(AlgebraicSymMatrix44()), 
+    theWeight(posTimeWeight), 
+    theWeightTimesPos(AlgebraicVector4()), 
+    theWeightInMix(weightInMix),
+    thePosAvailable(true), theTimeAvailable(true), theErrAvailable(false), theWeightAvailable(true), theWeightTimesPosAvailable(false),
+    valid(true), vertexIs4D(true)
+{}
+
+
+BasicSingleVertexState::BasicSingleVertexState(const AlgebraicVector4 & weightTimesPosition,
+                                               const GlobalWeight & posWeight,
+                                               const double & weightInMix)
+  : thePos(GlobalPoint(0, 0, 0)), 
+    theTime(dNaN), 
+    theErr(AlgebraicSymMatrix44()), 
     theWeight(posWeight), 
-    theWeightTimesPos(weightTimesPosition), 
+    theWeightTimesPos(weightTimesPosition),
     theWeightInMix(weightInMix),
-    thePosAvailable(false), theErrAvailable(false),theWeightAvailable(true), theWeightTimesPosAvailable(true),
-    valid(true)
+    thePosAvailable(false), theTimeAvailable(false), theErrAvailable(false), theWeightAvailable(true), theWeightTimesPosAvailable(true),
+    valid(true), vertexIs4D(true)
 {}
-   
 
 GlobalPoint BasicSingleVertexState::position() const
 {
-  if (!valid) throw VertexException("BasicSingleVertexState::invalid");
+  if (!valid) throw VertexException("BasicSingleVertexState::position::invalid");
   if (!thePosAvailable) computePosition();
   return thePos;
 }
 
-
 GlobalError BasicSingleVertexState::error() const
 {
-  if (!valid) throw VertexException("BasicSingleVertexState::invalid");
+  if (!valid) throw VertexException("BasicSingleVertexState::error::invalid");
+  if (!theErrAvailable) computeError();
+  return GlobalError(theErr.matrix());
+}
+
+GlobalError BasicSingleVertexState::error4D() const
+{
+  if (!valid) throw VertexException("BasicSingleVertexState::error4D::invalid");
   if (!theErrAvailable) computeError();
   return theErr;
 }
 
+double BasicSingleVertexState::time() const {
+  if ( vertexIs4D ) {
+    if (!valid) throw VertexException("BasicSingleVertexState::time::invalid");
+    if (!theTimeAvailable) computePosition(); // time computed with position (4-vector)
+    return theTime;
+  }
+  return dNaN;
```

Comment:
isn't theTime always dNaN in this case (whatever the default constructed value is)?
just "return theTime"  once at the end of the scope